shell/shell.go

```diff
@@ -57,3 +58,37 @@ func RmLink(last, path string) (string, error) {
 func AddLink(last, path, hash string) (string, error) {
 	return Exec("object", "patch", "add-link", last, path, hash)
 }
+
+func keyExists(key string) bool {
+	keys, err := Exec("key", "list", "-l")
+	if err != nil {
+		return false
+	}
+	lines := strings.Split(keys, "\n")
+	for _, line := range lines {
+		keyNames := strings.Split(line, " ")
+		for _, keyName := range keyNames {
+			if keyName != "" && keyName == key {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// Publish - Publishes the ipfs hash to the provided ipns key
+func Publish(cmd *exec.Cmd, hash, key string) (*exec.Cmd, error) {
```

Comment:
follow the pattern of other functions

Reply:
On second thought you can extend this API with `context.Context` to use [`WithTimeout`](https://golang.org/pkg/context/#WithTimeout).

Reply:
I tried various method including context and sending channels. All of them failed. The only approach that worked was cmd.Process.Kill(). The issue is that exec.Command is a blocking command and that's why context and channel approach didn't work here

Reply:
https://golang.org/pkg/os/exec/#CommandContext

Reply:
Yes but this is only based on timeout. We need a manual trigger whenever there is a new hash we have to kill the old process.

Reply:
In main you would create a channel that hashes are published to and start a goroutine that listens on this channel.
Hash received on this channel would execute a command in goroutine with context. Cancel function would be stored as local variable in goroutine. Whenever new hash is received `cancel()` is called.

Reply:
In fact you don't need to store `cancel()` just `defer cancel()` in a loop with `select{}`

Reply:
Okay will work on these structural changes tomorrow.